network/mesh.go

```diff
@@ -0,0 +1,276 @@
+// Copyright (C) 2019-2025 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/p2p/discovery/backoff"
+)
+
+const meshThreadInterval = time.Minute
+
+type meshStrategy interface {
+	meshThread()
+	start()
+	stop()
+}
+
+type genericMeshStrategy struct {
+	ctx                context.Context
+	meshUpdateRequests chan meshRequest
+	meshThreadInterval time.Duration
+	wg                 sync.WaitGroup
+	meshConfig
+}
+
+type meshConfig struct {
+	backoff        backoff.BackoffStrategy
+	netMesh        func() bool
+	peerStatReport func()
+	closer         func()
+}
+
+type hybridRelayMeshStrategy struct {
+	genericMeshStrategy
+}
+
+type meshStrategyOption func(*meshConfig)
+
+func withMeshExpJitterBackoff() meshStrategyOption {
+	return func(cfg *meshConfig) {
+		// Add exponential backoff with jitter to the mesh thread to handle new networks startup
+		// when no DNS or DHT peers are available.
+		// The parameters produce approximate the following delays (although they are random but the sequence give the idea):
+		// 2 2.4 4.6 9 20 19.5 28 24 14 14 35 60 60
```

Comment:
They go up _and_ down?

Reply:
I guess that is the jitter part?

Reply:
Yeah, I just expected the jitter to be small compared to the backoff so I was surprised to see something go from 28 down to 14.  I guess the jitter is probably selected uniformly at random from some constant, so at low numbers the jitter dominates the growth.